component/examples/stream_example.rs

```diff
@@ -0,0 +1,144 @@
+    ///
+    /// This would be an excellent place to also add a scopeguard with a defer_panic so that if the
+    /// component panics without a user caught irrecoverable error, a descriptive error message and/
+    /// or stacktrace can be forwarded to the supervisor, and resource cleanup can occur before the
+    /// existing component goes out of scope.
+    pub async fn listen(
+        tx_irrecoverable: Sender<anyhow::Error>,
+        rx_cancellation: oneshotReceiver<()>,
+    ) {
+        // Initialize the concrete type
+        let m_tcp = MockTcpStream::new();
+
+        loop {
+            let mut buf = [0; 10];
+            let _len = match m_tcp.mock_read(&mut buf) {
+                Ok(_) => {} // process
+                Err(_) => {
+                    let e = anyhow!("missing something required");
+                    tx_irrecoverable
+                        .send(e)
+                        .await
+                        .expect("Could not send irrecoverable signal.");
+                    wait_for_cancellation(rx_cancellation).await;
+                    return;
+                }
+            };
+        }
+    }
+}
+
+/// Wait for the cancellation signal in order to ensure that the message we sent to the
+/// supervisor was received before we return which causes the join handle to complete. If we
+/// were to return immediately, there would be no guarantee the message we send will be received.
+async fn wait_for_cancellation(rx_cancellation: oneshotReceiver<()>) {
+    loop {
+        tokio::select! {
+            _ = rx_cancellation => {
+                println!("terminating component task");
+                break;
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl Manageable for MockTcpStreamComponent {
+    #[allow(clippy::async_yields_async)]
+    /// The start function spawns a tokio task supplied with a function that
+    /// should be constantly running.
+    async fn start(
+        &self,
+        tx_irrecoverable: Sender<anyhow::Error>,
+        rx_cancellation: oneshotReceiver<()>,
+    ) -> tokio::task::JoinHandle<()> {
+        println!("starting component task");
+        let handle: JoinHandle<()> = tokio::spawn(Self::listen(tx_irrecoverable, rx_cancellation));
+        handle
+    }
+
+    /// Implement this function to log the error messages or take any task-specific action such as
+    /// closing a file or terminating children tasks.
+    fn handle_irrecoverable(&self, irrecoverable: IrrecoverableError) -> Result<(), anyhow::Error> {
+        println!("Received irrecoverable error: {}", irrecoverable);
+        Ok(())
+    }
+}
+
+#[tokio::main]
+pub async fn main() -> Result<(), anyhow::Error> {
+    // Create a component
+    let stream_component = MockTcpStreamComponent {};
+
+    // Create a supervisor for the component
+    let supervisor = Supervisor::new(stream_component);
+
+    // Spawn the supervisor to start the component and supervision.
+    let _ = match supervisor.spawn().await {
```

Comment:
I assume on our applications we'll want to wrap the `supervisor.spawn` on a task itself? Or otherwise on a big tokio select statement where we'll have all our components ?

Reply:
Once we spawn the task, we don't have to select, join, or do anything with the join handle. The supervisor takes care of that for us, and this function will run until you ctrl + c.

Reply:
I might be missing something, but practically to execute the underlying code of the `supervisor.spawn()` we have to call the `await` method which will actually block the current thread execution. So let's assume that I want to create a few supervisor components, to make those execute I have to call the `await` for them ,right? So I imagine something like this:

```
let supervisor_1 = Supervisor::new(stream_component);

supervisor_1.spawn().await;


let supervisor_2 = Supervisor::new(stream_component);

supervisor_2.spawn().await;

```

now, the `supervisor_2.spawn()` won't be able to execute unless the `supervisor_1.spawn().await` has finished. However, the `spawn` method calls the `run()` which is an infinite loop and will not allow us to continue the execution unless it has been exited. Apologies if I am getting this all wrong.

Reply:
Ah, you would need to call spawn in its own tokio task for each supervisor, we have something like this currently for our authority code. That is a really good question.